core/sync/interpreter_job.cpp

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "interpreter_job.hpp"
+
+#include "common/logger.hpp"
+#include "vm/interpreter/impl/interpreter_impl.hpp"
+
+namespace fc::sync {
+
+  namespace {
+    auto log() {
+      static common::Logger logger = common::createLogger("interpreter");
+      return logger.get();
+    }
+  }  // namespace
+
+  InterpreterJob::InterpreterJob(
+      std::shared_ptr<storage::PersistentBufferMap> kv_store,
+      std::shared_ptr<vm::interpreter::Interpreter> interpreter,
+      libp2p::protocol::Scheduler &scheduler,
+      ChainDb &chain_db,
+      IpfsStoragePtr ipld,
+      Callback callback)
+      : kv_store_(kv_store),
+        interpreter_(std::move(interpreter)),
+        scheduler_(scheduler),
+        chain_db_(chain_db),
+        ipld_(std::move(ipld)),
+        callback_(std::move(callback)),
+        result_{nullptr,
+                vm::interpreter::InterpreterError::kChainInconsistency} {
+    assert(callback_);
+  }
+
+  outcome::result<void> InterpreterJob::start(const TipsetKey &head) {
+    if (active_) {
+      log()->warn("current job ({} -> {}) is still active, cancelling it",
+                  status_.current_height,
+                  status_.target_height);
+      std::ignore = cancel();
+    }
+
+    OUTCOME_TRYA(target_head_, chain_db_.getTipsetByKey(head));
+    status_.target_height = target_head_->height();
+
+    const auto &hash = target_head_->key.hash();
+
+    // maybe already interpreted
+    OUTCOME_TRY(maybe_result,
```

Comment:
`getSavedResult` returns error if interpreter returned error, so `OUTCOME_TRY` will think we don't have cached result